src/mumble/PipeWire.cpp

```diff
@@ -0,0 +1,423 @@
+}
+
+const QList< audioDevice > PipeWireInputRegistrar::getDeviceChoices() {
+	return { audioDevice(QStringLiteral("Mono"), QString::number(1)) };
+}
+
+void PipeWireInputRegistrar::setDeviceChoice(const QVariant &choice, Settings &settings) {
+	settings.qsPipeWireInput = choice.toString();
+}
+
+bool PipeWireInputRegistrar::canEcho(EchoCancelOptionID, const QString &) const {
+	return false;
+}
+
+bool PipeWireInputRegistrar::isMicrophoneAccessDeniedByOS() {
+	return false;
+}
+
+PipeWireOutputRegistrar::PipeWireOutputRegistrar() : AudioOutputRegistrar(QStringLiteral("PipeWire"), 100) {
+}
+
+AudioOutput *PipeWireOutputRegistrar::create() {
+	return new PipeWireOutput;
+}
+
+const QList< audioDevice > PipeWireOutputRegistrar::getDeviceChoices() {
+	return { audioDevice(QStringLiteral("Mono"), QString::number(1)),
+			 audioDevice(QStringLiteral("Stereo"), QString::number(2)),
+			 audioDevice(QStringLiteral("2.1"), QString::number(3)),
+			 audioDevice(QStringLiteral("3.1"), QString::number(4)),
+			 audioDevice(QStringLiteral("5.1"), QString::number(6)),
+			 audioDevice(QStringLiteral("7.1"), QString::number(8)) };
+}
+
+void PipeWireOutputRegistrar::setDeviceChoice(const QVariant &choice, Settings &settings) {
+	settings.qsPipeWireOutput = choice.toString();
+}
+
+bool PipeWireOutputRegistrar::usesOutputDelay() const {
+	return false;
+}
+
+void PipeWireInit::initialize() {
+	pws.reset(new PipeWireSystem());
+
+	pws->m_lock.lock();
+	pws->m_waiter.wait(&pws->m_lock, 1000);
+	pws->m_lock.unlock();
+
+	if (pws->m_ok) {
+		inputRegistrar.reset(new PipeWireInputRegistrar());
+		outputRegistrar.reset(new PipeWireOutputRegistrar());
+	} else {
+		pws.reset();
+	}
+}
+
+void PipeWireInit::destroy() {
+	inputRegistrar.reset();
+	outputRegistrar.reset();
+	pws.reset();
+}
+
+static PipeWireInit jai;
+
+PipeWireSystem::PipeWireSystem() : m_ok(false), m_users(0) {
+	const QStringList names{ "libpipewire.so", "libpipewire-0.3.so" };
+
+	for (const auto &name : names) {
+		m_lib.setFileName(name);
+		if (m_lib.load()) {
+			break;
+		}
+	}
+
+	if (!m_lib.isLoaded()) {
+		return;
+	}
+
+	RESOLVE(pw_get_library_version);
+	RESOLVE(pw_init);
+	RESOLVE(pw_deinit);
+	RESOLVE(pw_main_loop_new);
+	RESOLVE(pw_main_loop_destroy);
+	RESOLVE(pw_main_loop_get_loop);
+	RESOLVE(pw_main_loop_run);
+	RESOLVE(pw_main_loop_quit);
+	RESOLVE(pw_properties_new);
+	RESOLVE(pw_stream_new_simple);
+	RESOLVE(pw_stream_destroy);
+	RESOLVE(pw_stream_connect);
+	RESOLVE(pw_stream_dequeue_buffer);
+	RESOLVE(pw_stream_queue_buffer);
+
+	qInfo("PipeWire %s from %s", pw_get_library_version(), qPrintable(m_lib.fileName()));
+
+	pw_init(nullptr, nullptr);
+
+	m_ok = true;
+}
+
+PipeWireSystem::~PipeWireSystem() {
+	if (m_ok) {
+		pw_deinit();
+	}
+}
+
+PipeWireEngine::PipeWireEngine(const char *category, const pw_stream_events *events, void *param)
+	: m_ok(false), m_loop(nullptr), m_stream(nullptr) {
+	if (!pws->isOk()) {
+		return;
+	}
+
+	m_loop = pws->pw_main_loop_new(nullptr);
```

Comment:
Shouldn't we be using `pw_thread_loop_new` for our purporses? Or are we implementing the threading from our side? If no, wouldn't pipewire be able to block the Mumble process?

Reply:
Done!